### ISSUE TITLE
`remotejdk_21` (Zulu 21.0.9) lacks the JDK-8341779 backport, causing Error Prone false positives

### ISSUE THREAD

**Issue by davido** [bazelbuild/rules_java#369]
## Summary

`remotejdk_21` is pinned to Azul Zulu 21.46.19 / OpenJDK 21.0.9+10, and
this pin is unchanged through the latest release (9.7.0,
`java/repositories.bzl`). This JDK build appears to lack the backport of
[JDK-8341779](https://bugs.openjdk.org/browse/JDK-8341779) (a redo of
[JDK-8225377](https://bugs.openjdk.org/browse/JDK-8225377)), which fixed
javac's reading of type-use annotations on type-variable bounds from
class files.

The practical consequence: recent Error Prone (bundled with the
java_tools shipped by rules_java 9.5.0+) produces **false positives on
documented Guava API usage** when the Java 21 toolchain is used. Since
[error-prone#5429](https://github.com/google/error-prone/pull/5429),
the `NullArgumentForNonNullParameter` check trusts `@Nullable` type-use
annotations on generic bounds — e.g. Guava's
`<T extends @Nullable Object>` on `Iterables.getFirst()`. When javac
cannot read that annotation from the class file, the check concludes
the parameter is non-null and rejects the documented usage:

```java
// canonical, documented Guava usage; javadoc: "defaultValue - the
// default value to return if the iterable is empty"
Iterables.getFirst(iterable, null);
```

```
error: [NullArgumentForNonNullParameter] Null is not permitted for this parameter.
    (see https://errorprone.info/bugpattern/NullArgumentForNonNullParameter)
```

This exact false positive is analyzed by the Error Prone maintainers in
[error-prone#5436](https://github.com/google/error-prone/issues/5436):
it requires a JDK 21 update release **containing the JDK-8341779
backport** (21.0.8+), and they explicitly note that vendors exercise
discretion over backports — **Temurin 21.0.9+10 contains it; the
pinned Zulu 21.0.9+10 empirically does not** (the false positive
reproduces with `remotejdk_21` and disappears with `remotejdk_25`,
where the fix is present natively).

## Impact

Any project that compiles with
`--java_language_version=21 --java_runtime_version=remotejdk_21` and
recent java_tools/Error Prone gets spurious
`NullArgumentForNonNullParameter` errors on correct code. Gerrit Code
Review's Java 21 CI verification broke on ~20 call sites of this
pattern after updating to rules_java 9.5.0, while the identical build
on the Java 25 toolchain is clean; we had to demote the check to a
warning as a workaround.

## Reproduction

```sh
# any source file containing: Iterables.getFirst(someIterable, null);
bazel build \
  --java_language_version=21 --java_runtime_version=remotejdk_21 \
  --tool_java_language_version=21 --tool_java_runtime_version=remotejdk_21 \
  //your:target   # with -Xep:NullArgumentForNonNullParameter:ERROR
```

The same target builds cleanly with the 25 toolchain, and per
error-prone#5436 also with Temurin 21.0.9+10.

## Requested change

Please bump the `remotejdk_21` pins to a Zulu 21 build that contains
the JDK-8341779 backport. Azul has already published newer builds:
`zulu21.50.19-ca-jdk21.0.11` and `zulu21.52.15-ca-jdk21.0.12` are
available for all platforms via the
[Azul metadata API](https://api.azul.com/metadata/v1/zulu/packages/?java_version=21).

## References

* https://github.com/google/error-prone/issues/5436 — the false
  positive, root-caused by the check's author
* https://github.com/google/error-prone/pull/5429 — the Error Prone
  change that started trusting bound annotations
* https://bugs.openjdk.org/browse/JDK-8341779 — the javac fix that
  needs to be present in the JDK running javac
* https://bugs.openjdk.org/browse/JDK-8225377 — original javac issue
* https://gerrit-review.googlesource.com/c/gerrit/+/615401 — Gerrit Code Review workaround

**Comment by hvadehra** [comment]
> Please bump the remotejdk_21 pins

A PR would be very welcome

**Comment by davido** [comment]
> A PR would be very welcome

Done: #370.

One thing worth attention in review: Azul changed the macOS tar.gz layout after 21.0.9 — the archives now contain an application bundle with the JDK under Contents/Home instead of a flat image (the Zulu 25 macOS archives are still flat). The two macOS entries' strip_prefix therefore descends into Contents/Home; a plain version bump alone breaks the macOS toolchains.

Verified against Gerrit's tree with --override_module: the --java_language_version=21 build goes from 88 NullArgumentForNonNullParameter false positives with Zulu 21.0.9 to zero with 21.0.12, confirming the JDK-8341779 backport is present.

**Comment by davido** [comment]
Verified the remaining failure: check_remote_jdk_configs_test now passes the hash and strip_prefix phases and fails only in the mirror check — all six new archives currently return 404 on mirror.bazel.build, e.g.

https://mirror.bazel.build/cdn.azul.com/zulu/bin/zulu21.52.15-ca-jdk21.0.12-linux_x64.tar.gz

I also double-checked ed021f9 against both archive generations locally: for the old 21.0.9 macOS archive the first-sorted /bin match is the root-level bin symlink (confirming your observation — the bundle under Contents/Home was always there), giving the flat strip_prefix; for 21.0.12 the only match is Contents/Home/bin/, giving the nested one. Both resolve to exactly the pins in this PR, so once the six cdn.azul.com URLs from java/repositories.bzl are mirrored — which I believe is an internal step on your side, as in a27004a — the test should be fully green with no further changes needed here.

Tiny aside from running the script on macOS: egrep '/bin(|/)$' is rejected by BSD grep ("empty (sub)expression"); GNU grep is fine, so CI is unaffected — just a heads-up for anyone running it locally on a Mac.

**Comment by davido** [comment]
@hvadehra — we're still hitting this false positive with `rules_java` 9.8.0's newer Error Prone. The 21.0.12 backport landed here is necessary but not sufficient: on JDK ≤ 21 the fix is opt-in via `-XDaddTypeAnnotationsToSymbol=true`, and Bazel's JavaBuilder wires Error Prone directly and never passes it. Filed it with a minimal reproducer: https://github.com/bazelbuild/bazel/issues/30743
